### PR TITLE
Bugfix in list_engine

### PR DESCRIPTION
**Pull request by aurghs**
Currently ``list_engines`` returns the list of all installed backend plus the list of the internal ones. For the internal ones, there is no check on the installed dependencies.
Now the registration of the internal backends is done by the backends only if the needed dependencies are installed.  

- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by alexamici** [comment]
Using `importlib.util.find_spec ` is very elegant but much less readable than a `try: except ModuleNotFoundError:` block, I suggest to convert all checks on imports to them.

File: xarray/backends/common.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import logging
 import time
 import traceback
+import typing as T
```
Comment (alexamici) [inline xarray/backends/common.py:4]: In _xarray_ the use is to do `from typing import Dict`.

File: xarray/backends/cfgrib_.py [hunk 2]
```diff
@@ -38,7 +43,6 @@ class CfGribDataStore(AbstractDataStore):
     """
 
     def __init__(self, filename, lock=None, **backend_kwargs):
-        import cfgrib
```
Comment (alexamici) [inline xarray/backends/cfgrib_.py:41]: If the import is later in the code I would suggest not to remove this additional import. Technically is not needed, but it is very confusing to rely on an import statement that is later in the code.

**Review by alexamici — CHANGES_REQUESTED** [review]

File: xarray/backends/cfgrib_.py (hunk 2 above)
Comment (aurghs) [reply xarray/backends/cfgrib_.py:41]: I agree

**Review by alexamici — APPROVED** [review]

**Comment by alexamici** [comment]
With this change `xr.list_engines()` returns a dictionary with all usable engines.

The implementation is forward compatible, so once backend are split off the main package newer backend registered via the plugin interface have precedence.

I'll merge it.

**Pull request merged**